### PR TITLE
Reference Manual: RPM's Philosophy

### PR DESCRIPTION
**Pull request by ffesti**
First brain dump.

### CODE REVIEW

File: docs/manual/philosophy.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+## Levels of control
+
+Most things in RPM can be configured by macros. This allows different
+actors to set them and overwrite what was set previously.
+
+Most basic are the defaults delivered by RPM upstream. Distributions
+are able to overwrite them by patching or providing new macro
+files. Packages can also ship macro files for other packages to use as
+BuildRequires. At last most build related macros can also be
```
Comment (nwalfield) [inline docs/manual/philosophy.md:34]: Awkward: s/At last/Finally/ ?

File: docs/manual/philosophy.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+Nosource packages are a exception that does not ship the sources but
+only the build instructions to keep proprietary source private. They
+still can be used to generate binary packages in combination with the
+sources.
+
+#### Separate Upstream Source from Patches
+
+Source packages are supposed to contain the unaltered upstream sources
+in a way their checksum can be checked. All changes can be done by
```
Comment (nwalfield) [inline docs/manual/philosophy.md:102]: Copy edit: in a way *that*

File: docs/manual/philosophy.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+applying patches or running build scripts. This makes it easy to
+understand what the packager did to the software. This is important to
+be able to figure out where issue arise from and to see which issues
+are fixed in the package - even if the upstream version may still be
```
Comment (nwalfield) [inline docs/manual/philosophy.md:106]: Copy edit: s/may still be/is still/

File: docs/manual/philosophy.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+vulnerable.
+
+#### Support back porting fixes
+
+Distribution often have a different approach to fixing software than
```
Comment (nwalfield) [inline docs/manual/philosophy.md:111]: Copy edit: *A* distribution often *has* a different approach to fixing software *from an* upstream project.

File: docs/manual/philosophy.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+upstream project have. Upstream project are often content with fixing
```
Comment (nwalfield) [inline docs/manual/philosophy.md:112]: Copy edit: s/project/projects/

File: docs/manual/philosophy.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+their latest release(s). Distribution often support a particular
+version of a software and need those to be fixed. The Sources and
+Patches approach make handling this a lot easier (although it is
```
Comment (nwalfield) [inline docs/manual/philosophy.md:115]: Copy edit: s/make/makes/

File: docs/manual/philosophy.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+lately being extended with the use of version control like git). 
+
+### Allow 3rd party packages
+
+Although RPM is designed to package a whole distribution it explicitly
+supports 3rd parties to also provide packages without becoming part of
+the distribution and their processes.
+
+Rpmbuild can be run locally without the use of a big build
+system. Dependencies are in large parts automatic and target libraries
```
Comment (nwalfield) [inline docs/manual/philosophy.md:125]: Copy edit: s/parts/part/

File: docs/manual/philosophy.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+Sources are given as files and patches. Upstream tarballs should be
+used unchanged and patched during the build. This way packages can be
+reviewed more easily and the changes made to the upstream project are
+explicit.
+
+
+## Handling all system files
+
+RPM takes owner ship of all system files and their live cycle. While
```
Comment (nwalfield) [inline docs/manual/philosophy.md:140]: Copy edit: s/owner ship/ownership/, s/live/life/

File: docs/manual/philosophy.md [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+packages can copy files in their scriptlets at installation time this
+is strongly discouraged. Even files that are not actually shipped can
+be added to the package as %ghost files to allow RPM to handle
+them. All system files and directories should be owned by a package.
+
+Scriptlets should be used sparingly. Most use cases - like updating
+caches and indexes - can be dealt with a central filetrigger. Although
```
Comment (nwalfield) [inline docs/manual/philosophy.md:147]: Awkard: s/with a/with by using a/ ?

**Comment by ffesti** [comment]
Thanks a lot! Good to have a native speaker reading over this. Added the suggested changes.

File: docs/manual/philosophy.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
+---
+layout: default
+title: rpm.org - RPM's Philosophy
+---
+# RPM's Philosophy
+
+The RPM package manager is a general purpose software manager. It
```
Comment (pmatilai) [inline docs/manual/philosophy.md:7]: I think a brief definition of what that means would be in order here, something that covers the two sides of our operation: wrapping software into packages that the other side knows how to manipulate: install, update, remove, verify.

Comment (pmatilai) [reply docs/manual/philosophy.md:7]: I'd also place a summary of the founding design principles right here in the beginning to set a frame of mind:
- unattended operation
- reproducibility
- verifiability

File: docs/manual/philosophy.md [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+Libraries should be packaged separately and binaries should link to
+the version provided by system packages. Static linking is
+discouraged. This limits the number of packages that need updates or
+re-builds in case of a vulnerability.
+
+#### Unattended operation
+
+Package installation and update is unattended and must not require
+user interaction. When possible delay user interaction to the first
+start up.
```
Comment (pmatilai) [inline docs/manual/philosophy.md:77]: Just drop the second sentence, user interaction is simply not an option during operation.

File: docs/manual/philosophy.md [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+inside the packages instead of relying on conventions like package
+names. This way 3rd party packages can be checked for compatibility
+beyond them just claiming that they target the distribution in question.
+
+## Separating Upstream Sources and Packager's Changes
+
+Sources are given as files and patches. Upstream tarballs should be
+used unchanged and patched during the build. This way packages can be
+reviewed more easily and the changes made to the upstream project are
+explicit.
```
Comment (pmatilai) [inline docs/manual/philosophy.md:135]: The rpm lore term for this is "pristine sources", I'd stick it somewhere in there :sweat_smile:

Comment (pmatilai) [reply docs/manual/philosophy.md:135]: Also this is practically the same as "Separate Upstream Source from Patches" a couple of sections earlier. I get that there's a slight difference in perspective, but  its quite repetitive. Backporting fixies too. 

Maybe have a single "Pristine sources" section for the main beef, and elaborate on these things it enables as subsections?

File: docs/manual/about.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
+---
+layout: default
+title: rpm.org - About the Reference Manual
+---
+# About this manual
+
+This is the Refence Manual for the RPM Package Manager. Unfortunately
+it is still incomplete. Recently added features are described with
+decent detail but more basic properties are only being added step by
+step.
```
Comment (pmatilai) [inline docs/manual/about.md:10]: This is a much needed disclaimer indeed :smile: :+1:

File: docs/manual/philosophy.md [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+---
+# RPM's Philosophy
+
+The RPM package manager is a general purpose software manager. It
+differs from special purpose package managers - like those targeting a
+specific programming language - in trying to not make assumptions about
+how software looks or gets packaged. Packaging software can be messy and
+RPM accomodates for that.
+
+It still offers help to create a POSIX like operation system. Marcos
```
Comment (pmatilai) [inline docs/manual/philosophy.md:13]: Maybe "help in creating a ..."?

Comment (pmatilai) [reply docs/manual/philosophy.md:13]: Oh, and s/Marcos/Macros/

File: docs/manual/philosophy.md (hunk 11 above)
Comment (pmatilai) [reply docs/manual/philosophy.md:77]: Builds being unattended too doesn't seem to be mentioned anywhere, that's an equally important thing.

**Comment by pmatilai** [comment]
Lots of useful stuff here. I'd maybe move the Design goals as the first subsection so it goes from higher level towards details. Jumping straight into talking about macros seems like getting ahead of ourselves.

Finally, these design goals appear to be missing entirely:
- reproducibility (of builds and installs)
- verifiability (of packages and installed software)

**Comment by pmatilai** [comment]
Just realized this will make a nice addition to the contributing guide: when contributing / planning to, make sure the work aligns with rpm's link-to-design-philosophy. Doesn't have to be added in this very PR but wouldn't hurt either.

**Comment by ffesti** [comment]
Ok, I don't have the mental capacity to make this into a corner stone of western literature but it should now pass as a piece of our docs.

File: docs/manual/philosophy.md [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+As such RPM is only a part in a larger set of tools and some users may
+never actually interact with RPM directly.
+
+## Design goals
+
+* General purpose software management
+* Building from pristine sources
+* Unattended operation
+  * for both building
```
Comment (pmatilai) [inline docs/manual/philosophy.md:58]: "for both building" doesn't make sense, looks like an editing leftover :)

Comment (ffesti) [reply docs/manual/philosophy.md]: Fixed

**Pull request merged**